prusti-interface/src/environment/mod.rs

```diff
@@ -277,36 +287,96 @@ impl<'tcx> Environment<'tcx> {
+        // SAFETY: This is safe because we are feeding in the same `tcx`
+        // that was used to store the data.
+        let body_with_facts = unsafe {
+            self::mir_storage::retrieve_mir_body(self.tcx, def_id)
+        };
+
+        let facts = BorrowckFacts {
+            input_facts: RefCell::new(Some(body_with_facts.input_facts)),
+            output_facts: body_with_facts.output_facts,
+            location_table: RefCell::new(Some(body_with_facts.location_table)),
+        };
+
+        CachedLocalBody {
+            body: CachedBody {
+                base_body: Rc::new(body_with_facts.body),
+                monomorphised_bodies: HashMap::new(),
+            },
+            borrowck_facts: Rc::new(facts),
+        }
+    }
+
     /// Get the MIR body of a local procedure, monomorphised with the given
     /// type substitutions.
-    pub fn local_mir(
-        &self,
-        def_id: LocalDefId,
-        substs: SubstsRef<'tcx>,
-    ) -> Rc<mir::Body<'tcx>> {
+    pub fn local_body_mir(&self, def_id: LocalDefId, substs: SubstsRef<'tcx>) -> Rc<mir::Body<'tcx>> {
+        // look first into the map of pre-fetched pure functions
+        let mut pure_function_bodies = self.pure_function_bodies.borrow_mut();
+        let mut predicate_bodies = self.predicate_bodies.borrow_mut();
         let mut bodies = self.bodies.borrow_mut();
-        let body = bodies.entry(def_id)
-            .or_insert_with(|| {
-                // SAFETY: This is safe because we are feeding in the same `tcx`
-                // that was used to store the data.
-                let body_with_facts = unsafe {
-                    self::mir_storage::retrieve_mir_body(self.tcx, def_id)
-                };
-                let body = body_with_facts.body;
-                let facts = BorrowckFacts {
-                    input_facts: RefCell::new(Some(body_with_facts.input_facts)),
-                    output_facts: body_with_facts.output_facts,
-                    location_table: RefCell::new(Some(body_with_facts.location_table)),
-                };
-
-                CachedBody {
-                    base_body: Rc::new(body),
-                    monomorphised_bodies: HashMap::new(),
-                    borrowck_facts: Rc::new(facts),
-                }
-            });
-        body
-            .monomorphised_bodies
+        let body = if let Some(body) = pure_function_bodies.get_mut(&def_id) {
+            body
+        } else if let Some(body) = predicate_bodies.get_mut(&def_id) {
+            body
+        } else {
+            // fallback to fetching the non-pure mir body, which may not be cached yet
+            let body = bodies.entry(def_id)
+                .or_insert_with(|| {
+                    self.raw_load_local_mir(def_id)
+                });
+            body
+        };
+
+        self.subst_into_body(&mut body.body, substs)
+    }
+
+    // used only internally
```

Comment:
This is already implied by the visibility, but what does the function actually do?